 ##  Phishing Website Detection Feature Extraction

#### first we have import pandas

In [2]:
import pandas as pd

##### then we have install and import weget module for loading data from phishtank portal

In [2]:
! pip install wget
import wget

  Using cached wget-3.2-py3-none-any.whl


###### and we have dowload data from phishtank and save in dowloaded_file.csv file

In [5]:
url = 'http://data.phishtank.com/data/online-valid.csv'
output_filename = 'downloaded_file1.csv'

wget.download(url, out=output_filename)

HTTPError: HTTP Error 429: Too Many Requests

Now we load our dowload file 

In [3]:
data0 = pd.read_csv("Data/downloaded_file.csv")
data0.head()

phish_id                                                url  \
0   8449707                        https://dobe.page.link/qbvQ   
1   8449706  https://wpw.xvx.mybluehost.me/website_f6cfd0e0...   
2   8449705  https://wpw.xvx.mybluehost.me/website_f6cfd0e0...   
3   8449700             https://swisspasschpbjectives.web.app/   
4   8449698  https://pub-62151b6ea63b48fd8aede30a0de6256a.r...   

                                    phish_detail_url  \
0  http://www.phishtank.com/phish_detail.php?phis...   
1  http://www.phishtank.com/phish_detail.php?phis...   
2  http://www.phishtank.com/phish_detail.php?phis...   
3  http://www.phishtank.com/phish_detail.php?phis...   
4  http://www.phishtank.com/phish_detail.php?phis...   

             submission_time verified          verification_time online target  
0  2024-02-12T10:09:53+00:00      yes  2024-02-12T10:12:16+00:00    yes  Other  
1  2024-02-12T10:09:38+00:00      yes  2024-02-12T10:12:16+00:00    yes  Other  
2  2024-02-12T10:09:37+00:00      yes  2024-02-12T10:12:16+00:00    yes  Other  
3  2024-02-12T10:09:14+00:00      yes  2024-02-12T10:12:16+00:00    yes  Other  
4  2024-02-12T10:08:56+00:00      yes  2024-02-12T10:12:16+00:00    yes  Other

In [4]:
data0.shape

(46647, 8)

now we have randomly select 5000 row form above dataframe because above dataframe have 46647 row .tahe data is updated hourly i am considering 5000 urls phishing and 5000 url legitimate


In [5]:
phishurl = data0.sample(n = 5000, random_state = 12).copy()
phishurl = phishurl.reset_index(drop=True)
phishurl.head()

phish_id                                                url  \
0   8330256  https://ipfs.eth.aragon.network/ipfs/bafybeiar...   
1   8248695  https://cloudflare-ipfs.com/ipfs/bafybeihfnmzg...   
2   8187964  https://cf-ipfs.com/ipfs/QmRfmvPBectE5FTCqf65T...   
3   8364764  https://cloudflare-ipfs.com/ipfs/bafybeic6xofh...   
4   8262243  https://forms.office.com/pages/responsepage.as...   

                                    phish_detail_url  \
0  http://www.phishtank.com/phish_detail.php?phis...   
1  http://www.phishtank.com/phish_detail.php?phis...   
2  http://www.phishtank.com/phish_detail.php?phis...   
3  http://www.phishtank.com/phish_detail.php?phis...   
4  http://www.phishtank.com/phish_detail.php?phis...   

             submission_time verified          verification_time online target  
0  2023-10-12T23:14:22+00:00      yes  2023-10-12T23:34:19+00:00    yes  Other  
1  2023-08-03T22:39:09+00:00      yes  2023-08-03T22:53:11+00:00    yes  Other  
2  2023-06-20T04:44:39+00:00      yes  2023-06-20T04:53:02+00:00    yes  Other  
3  2023-11-15T05:52:22+00:00      yes  2023-11-15T06:09:35+00:00    yes  Other  
4  2023-08-16T22:11:01+00:00      yes  2023-08-16T22:32:20+00:00    yes  Other

In [6]:
phishurl.shape

(5000, 8)

now we have load our legitimate url which is dowlaaded from https://www.unb.ca/cic/datasets/index.html 

In [7]:
data1 = pd.read_csv("Data/Benign_list_big_final.csv")
data1.columns = ['URLs']
data1.head()

URLs
0  http://1337x.to/torrent/1110018/Blackhat-2015-...
1  http://1337x.to/torrent/1122940/Blackhat-2015-...
2  http://1337x.to/torrent/1124395/Fast-and-Furio...
3  http://1337x.to/torrent/1145504/Avengers-Age-o...
4  http://1337x.to/torrent/1160078/Avengers-age-o...

now we have 35300 legitimate  urls

and we have rondomly select 5000 urls 

In [8]:
legiurl = data1.sample(n = 5000, random_state = 12).copy()
legiurl = legiurl.reset_index(drop=True)
legiurl.head()

URLs
0  http://graphicriver.net/search?date=this-month...
1  http://ecnavi.jp/redirect/?url=http://www.cros...
2  https://hubpages.com/signin?explain=follow+Hub...
3  http://extratorrent.cc/torrent/4190536/AOMEI+B...
4  http://icicibank.com/Personal-Banking/offers/o...

In [9]:
legiurl.shape

(5000, 1)

now we have import some important module which important while working with url

In [10]:
from urllib.parse import urlparse,urlencode
import ipaddress
import re

## Feature Extraction



1. Address Bar based Features
2. Domain based Features
3. HTML & Javascript based Features

In [11]:
# 1.Domain of the URL (Domain) 
def getDomain(url):  
  domain = urlparse(url).netloc
  if re.match(r"^www.",domain):
       domain = domain.replace("www.","")
  return domain

In [12]:
# 2.Checks for IP address in URL (Have_IP)
def havingIP(url):
  try:
    ipaddress.ip_address(url)
    ip = 1
  except:
    ip = 0
  return ip

In [13]:
# 3.Checks the presence of @ in URL (Have_At)
def haveAtSign(url):
  if "@" in url:
    at = 1    
  else:
    at = 0    
  return at

In [14]:
# 4.Finding the length of URL and categorizing (URL_Length)
def getLength(url):
  if len(url) < 54:
    length = 0            
  else:
    length = 1            
  return length

In [15]:
# 5.Gives number of '/' in URL (URL_Depth)
def getDepth(url):
  s = urlparse(url).path.split('/')
  depth = 0
  for j in range(len(s)):
    if len(s[j]) != 0:
      depth = depth+1
  return depth

In [16]:
# 6.Checking for redirection '//' in the url (Redirection)
def redirection(url):
  pos = url.rfind('//')
  if pos > 6:
    if pos > 7:
      return 1
    else:
      return 0
  else:
    return 0

In [17]:
# 7.Existence of “HTTPS” Token in the Domain Part of the URL (https_Domain)
def httpDomain(url):
  domain = urlparse(url).netloc
  if 'https' in domain:
    return 1
  else:
    return 0

In [18]:
#listing shortening services
shortening_services = r"bit\.ly|goo\.gl|shorte\.st|go2l\.ink|x\.co|ow\.ly|t\.co|tinyurl|tr\.im|is\.gd|cli\.gs|" \
                      r"yfrog\.com|migre\.me|ff\.im|tiny\.cc|url4\.eu|twit\.ac|su\.pr|twurl\.nl|snipurl\.com|" \
                      r"short\.to|BudURL\.com|ping\.fm|post\.ly|Just\.as|bkite\.com|snipr\.com|fic\.kr|loopt\.us|" \
                      r"doiop\.com|short\.ie|kl\.am|wp\.me|rubyurl\.com|om\.ly|to\.ly|bit\.do|t\.co|lnkd\.in|db\.tt|" \
                      r"qr\.ae|adf\.ly|goo\.gl|bitly\.com|cur\.lv|tinyurl\.com|ow\.ly|bit\.ly|ity\.im|q\.gs|is\.gd|" \
                      r"po\.st|bc\.vc|twitthis\.com|u\.to|j\.mp|buzurl\.com|cutt\.us|u\.bb|yourls\.org|x\.co|" \
                      r"prettylinkpro\.com|scrnch\.me|filoops\.info|vzturl\.com|qr\.net|1url\.com|tweez\.me|v\.gd|" \
                      r"tr\.im|link\.zip\.net"

In [19]:
# 8. Checking for Shortening Services in URL (Tiny_URL)
def tinyURL(url):
    match=re.search(shortening_services,url)
    if match:
        return 1
    else:
        return 0

In [20]:
# 9.Checking for Prefix or Suffix Separated by (-) in the Domain (Prefix/Suffix)
def prefixSuffix(url):
    if '-' in urlparse(url).netloc:
        return 1            # phishing
    else:
        return 0            # legitimate

In [21]:
!pip install python-whois

In [22]:
# importing required packages for this section
import re
from bs4 import BeautifulSoup
import whois
import urllib
import urllib.request
from datetime import datetime

In [23]:
# 13.Survival time of domain: The difference between termination time and creation time (Domain_Age)  
def domainAge(domain_name):
  creation_date = domain_name.creation_date
  expiration_date = domain_name.expiration_date
  if (isinstance(creation_date,str) or isinstance(expiration_date,str)):
    try:
      creation_date = datetime.strptime(creation_date,'%Y-%m-%d')
      expiration_date = datetime.strptime(expiration_date,"%Y-%m-%d")
    except:
      return 1
  if ((expiration_date is None) or (creation_date is None)):
      return 1
  elif ((type(expiration_date) is list) or (type(creation_date) is list)):
      return 1
  else:
    ageofdomain = abs((expiration_date - creation_date).days)
    if ((ageofdomain/30) < 6):
      age = 1
    else:
      age = 0
  return age

In [24]:
# 14.End time of domain: The difference between termination time and current time (Domain_End) 
def domainEnd(domain_name):
  expiration_date = domain_name.expiration_date
  if isinstance(expiration_date,str):
    try:
      expiration_date = datetime.strptime(expiration_date,"%Y-%m-%d")
    except:
      return 1
  if (expiration_date is None):
      return 1
  elif (type(expiration_date) is list):
      return 1
  else:
    today = datetime.now()
    end = abs((expiration_date - today).days)
    if ((end/30) < 6):
      end = 0
    else:
      end = 1
  return end

In [25]:
# importing required packages for this section
import requests

In [26]:
# 15. IFrame Redirection (iFrame)
def iframe(response):
  if response == "":
      return 1
  else:
      if re.findall(r"[<iframe>|<frameBorder>]", response.text):
          return 0
      else:
          return 1

In [27]:
# 16.Checks the effect of mouse over on status bar (Mouse_Over)
def mouseOver(response): 
  if response == "" :
    return 1
  else:
    if re.findall("<script>.+onmouseover.+</script>", response.text):
      return 1
    else:
      return 0

In [28]:
# 17.Checks the status of the right click attribute (Right_Click)
def rightClick(response):
  if response == "":
    return 1
  else:
    if re.findall(r"event.button ?== ?2", response.text):
      return 0
    else:
      return 1

In [29]:
# 18.Checks the number of forwardings (Web_Forwards)    
def forwarding(response):
  if response == "":
    return 1
  else:
    if len(response.history) <= 2:
      return 0
    else:
      return 1

In [30]:
#Function to extract features
def featureExtraction(url,label):

  features = []
  #Address bar based features (10)
  features.append(getDomain(url))
  features.append(havingIP(url))
  features.append(haveAtSign(url))
  features.append(getLength(url))
  features.append(getDepth(url))
  features.append(redirection(url))
  features.append(httpDomain(url))
  features.append(tinyURL(url))
  features.append(prefixSuffix(url))
  
  #Domain based features (4)
  dns = 0
  try:
    domain_name = whois.whois(urlparse(url).netloc)
  except:
    dns = 1

  features.append(dns)
  #features.append(web_traffic(url))
  features.append(1 if dns == 1 else domainAge(domain_name))
  features.append(1 if dns == 1 else domainEnd(domain_name))
  
  # HTML & Javascript based features (4)
  try:
    response = requests.get(url)
  except:
    response = ""
  features.append(iframe(response))
  features.append(mouseOver(response))
  features.append(rightClick(response))
  features.append(forwarding(response))
  features.append(label)
  
  return features

In [31]:
legiurl.shape

(5000, 1)

In [ ]:
#Extracting the feautres & storing them in a list
legi_features = []
label = 0

for i in range(0, 100):
  url = legiurl['URLs'][i]
  legi_features.append(featureExtraction(url,label))

2024-05-19 14:54:21,310 - whois.whois - ERROR - Error trying to connect to socket: closing socket - [Errno 11001] getaddrinfo failed


In [ ]:
print(legi_features.shape)

In [ ]:
#converting the list to dataframe
feature_names = ['Domain', 'Have_IP', 'Have_At', 'URL_Length', 'URL_Depth','Redirection', 
                      'https_Domain', 'TinyURL', 'Prefix/Suffix', 'DNS_Record', 
                      'Domain_Age', 'Domain_End', 'iFrame', 'Mouse_Over','Right_Click', 'Web_Forwards', 'Label']

legitimate = pd.DataFrame(legi_features, columns= feature_names)
legitimate.head()

In [ ]:
# Storing the extracted legitimate URLs fatures to csv file
legitimate.to_csv('legitimate.csv', index= False)

In [ ]:
phishurl.shape

In [ ]:
#Extracting the feautres & storing them in a list
phish_features = []
label = 1
for i in range(0, 5000):
  url = phishurl['url'][i]
  phish_features.append(featureExtraction(url,label))

In [ ]:
#converting the list to dataframe
feature_names = ['Domain', 'Have_IP', 'Have_At', 'URL_Length', 'URL_Depth','Redirection', 
                      'https_Domain', 'TinyURL', 'Prefix/Suffix', 'DNS_Record', 
                      'Domain_Age', 'Domain_End', 'iFrame', 'Mouse_Over','Right_Click', 'Web_Forwards', 'Label']

phishing = pd.DataFrame(phish_features, columns= feature_names)
phishing.head()

In [ ]:
# Storing the extracted legitimate URLs fatures to csv file
phishing.to_csv('phishing.csv', index= False)

In [ ]:
#Concatenating the dataframes into one 
urldata = pd.concat([legitimate, phishing]).reset_index(drop=True)
urldata.head()

In [ ]:
urldata.tail()

In [ ]:
urldata.shape

In [ ]:
# Storing the data in CSV file
urldata.to_csv('urldata.csv', index=False)